File: ROS/ROS_21_Tutorials/catkin_ws/src/learning_parameter/src/handleParameter.cpp
/**
 * @file    handleParameter.cpp
 * @brief   设置/读取ROS参数的例程
 * @details 使用get/set函数来读取/设置ROS参数
 * @author  Jack Zhang
 * @email   zhangmengwei1996@outlook.com
 * @version 1.0.0
 * @date    3/11/21
 * @license GNU General Public License (GPL)
 */
#include <ros/ros.h>
#include <std_srvs/Empty.h>

int main(int argc, char **argv)
{
	ros::init(argc, argv, "handleParameter"); // 初始化ROS节点
	ros::NodeHandle nodeHandle; // 创建节点句柄

	int red, green, blue;

	// 读取背景颜色的参数到变量
	ros::param::get("/turtlesim/background_r", std::ref(red));
	ros::param::get("/turtlesim/background_g", std::ref(green));
	ros::param::get("/turtlesim/background_b", std::ref(blue));
	ROS_INFO("Get background color: [%d, %d, %d]", red, green, blue);

	// 设置背景颜色参数
	ros::param::set("/turtlesim/background_r", 100);
	ros::param::set("/turtlesim/background_g", 100);
	ros::param::set("/turtlesim/background_b", 100);
	ROS_INFO("Set background color: [100, 100, 100 ]");

	// 读取背景颜色的参数到变量
	ros::param::get("/turtlesim/background_r", std::ref(red));
	ros::param::get("/turtlesim/background_g", std::ref(green));
	ros::param::get("/turtlesim/background_b", std::ref(blue));
	ROS_INFO("Get background color: [%d, %d, %d]", red, green, blue);

	ros::service::waitForService("/clear");
	ros::ServiceClient clearBackground = nodeHandle.serviceClient<std_srvs::Empty>("/clear");
	// 调用服务，刷新背景颜色
	std_srvs::Empty srvEmpty;
	clearBackground.call(srvEmpty);

	sleep(1);
	return 0;
}

File: ROS/ROS_21_Tutorials/catkin_ws/src/learning_parameter/scripts/handle_parameter.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# 该例程设置/读取ROS参数
import rospy
from std_srvs.srv import Empty
import random


def handle_parameter():
    rospy.init_node("handle_parameter")  # ROS节点初始化

    # 读取背景颜色参数
    red = rospy.get_param("/turtlesim/background_r")
    green = rospy.get_param("/turtlesim/background_g")
    blue = rospy.get_param("/turtlesim/background_b")
    rospy.loginfo("Got turtlesim background color: [{}, {}, {}]".format(red, green, blue))

    # 设置背景颜色参数
    rand_color = int(random.random() * 255)
    rospy.set_param("/turtlesim/background_r", rand_color)
    rospy.set_param("/turtlesim/background_g", rand_color)
    rospy.set_param("/turtlesim/background_b", rand_color)
    rospy.loginfo("Set turtlesim background color: [{}, {}, {}]".format(rand_color, rand_color, rand_color))

    # 读取背景颜色参数
    red = rospy.get_param("/turtlesim/background_r")
    green = rospy.get_param("/turtlesim/background_g")
    blue = rospy.get_param("/turtlesim/background_b")
    rospy.loginfo("Got turtlesim background color: [{}, {}, {}]".format(red, green, blue))

    # 阻塞，等待直到发现/clear服务
    rospy.wait_for_service("/clear")

    try:
        clear_background = rospy.ServiceProxy("/clear", Empty)
        response = clear_background()
        return response
    except rospy.ServiceException as e:
        print("Service call failed: %s" % e)


if __name__ == "__main__":
    handle_parameter()
